main/src/library/Fixpoint3/PrecedenceGraph.flix

```diff
@@ -0,0 +1,322 @@
+/// of a Datalog program. The stratification is based on a topological sort
+/// of the strongly connected components (SCC) of the precedence graph.
+///
+/// The topological sort described a legal stratification of the Datalog program.
+///
+/// A `PseudoSCC` is a group of mutually independent, consecutive SCC in the topological
+/// order.
+///
+mod Fixpoint3.PrecedenceGraph {
+    @Internal
+    pub type alias Vertex = Int32
+
+    ///
+    /// A `PseudoSCC` is a list of `SCC`s without dependencies between them.
+    ///
+    @Internal
+    pub type alias PseudoSCC = List[Set[Vertex]]
+
+    ///
+    /// Returns a list of the pseudo-SCCs of `g`.
+    ///
+    pub def getPseudoStrata(g: MutGraph[r]): List[PseudoSCC] \ r = region rc {
+        let sorted = MutGraph.getSCCOrder(g);
+        let pseudoStratums = MutList.empty(rc);
+        def loop(s1, list) = match list {
+            case s2 :: rest =>
+                let mergable = s1 |>
+                    List.forAll(
+                        Set.forAll(u -> s2 |> Set.forAll(v -> not MutGraph.hasEdge(u, v, g)))
+                    );
+                let next =
+                    if (mergable) s2 :: s1
+                    else {
+                        MutList.push(s1, pseudoStratums);
+                        s2 :: Nil
+                    };
+                loop(next, rest)
+            case Nil => MutList.push(s1, pseudoStratums)
+        };
+        match sorted {
+            case s :: rest => loop(s :: Nil, rest)
+            case Nil => ()
+        };
+        pseudoStratums |> MutList.toList
+    }
+
+    @Internal
+    pub struct MutGraph[r] {
+        vertices: MutSet[Vertex, r],
+        adjList: MutMap[Vertex, List[Vertex], r]
+    }
+
+    mod MutGraph {
+        use Fixpoint3.Counter
+        use Fixpoint3.PrecedenceGraph.Vertex
+        use Fixpoint3.Util.getOrCrash
+
+        ///
+        /// `Color` is used for computing the topological sort.
+        ///
+        enum Color with Eq {
+            case White,
+            case Gray,
+            case Black
+        }
+
+        ///
+        /// Adds the edge `(u, v)` to `g`.
+        ///
+        @Internal
+        pub def addEdge(u: Vertex, v: Vertex, g: MutGraph[r]): Unit \ r = {
+            MutSet.add(v, g->vertices);
+            MutSet.add(u, g->vertices);
+            let neighbors = MutMap.getWithDefault(u, List.empty(), g->adjList);
+            if (List.memberOf(v, neighbors)) ()
+            else MutMap.put(u, v :: neighbors, g->adjList)
+        }
+
+        ///
+        /// Adds the edge `(u, v)` to `g`.
+        ///
+        @Internal
+        pub def addVertex(v: Vertex, g: MutGraph[r]): Unit \ r = MutSet.add(v, g->vertices)
+
+        ///
+        /// Returns `true` if `g` contains the edge `(u, v)`.
+        ///
+        @Internal
+        pub def hasEdge(u: Vertex, v: Vertex, g: MutGraph[r]): Bool \ r =
+            g->adjList |> MutMap.getWithDefault(u, List.empty()) |> List.memberOf(v)
```

Comment:
I was a bit surprised to see List.memberOf here, I'm guessing that having adjList be a set is slower since you dont do it?